### PR TITLE
feat(cli): support ephemeral parameters

### PR DESCRIPTION
**Pull request by mtojek**
Related: https://github.com/coder/coder/issues/6828

This PR extends CLI to support _one-time parameters/ephemeral parameters/build_ options.

Screenshot 2023-07-11 at 16 35 29

The flag is available for:
* `coder update ...`
* `coder create ...`
* `coder start ...`
* `coder restart ...`

_I can rename the flag name  (`--build-options` ) to something more meaningful if anyone has an idea._

---

TODO:
- [x] `update` and `create`: skip ephemeral parameters unless the workspace user wants to set them (flag `--build-options`)
- [x] ... with unit tests
- [x] `start`: add `--build-options` flag to ask for ephemeral parameters
- [x] ... with unit tests
- [x] `restart`: add `--build-options` flag to ask for ephemeral parameters
- [x] ... with unit tests

### CODE REVIEW

File: cli/create.go [hunk 1]
```diff
@@ -189,6 +191,11 @@ func (r *RootCmd) create() *clibase.Cmd {
 			Description: "Specify a duration after which the workspace should shut down (e.g. 8h).",
 			Value:       clibase.DurationOf(&stopAfter),
 		},
+		clibase.Option{
+			Flag:        "build-options",
+			Description: "Prompt for one-time build options defined with ephemeral parameters.",
+			Value:       clibase.BoolOf(&buildOptions),
+		},
```
Comment (johnstcn) [inline cli/create.go:198]: What's the rationale behind this? I thought we would always prompt for ephemeral parameters? In this case, if we do not speciify `--build-options` and the template defines ephemeral parameters, would the build not just fail?

Comment (mtojek) [reply cli/create.go:198]: From a user perspective, they are [optional](https://github.com/coder/coder/issues/6828#issuecomment-1623843335). Workspace owner can either start the workspace or start with options (= ephemeral parameters). It is expected not to ask for ephemeral values every time.

> What's the rationale behind this? I thought we would always prompt for ephemeral parameters? In this case, if we do not speciify --build-options and the template defines ephemeral parameters, would the build not just fail?

They are still parameters, so Coder will switch to default values for them. If the user defines `default: ""`, then Coder will assume an empty string.

Ephemerality just prevents Coder from passing values between workspace builds. Everything else works for parameters the same way.

Comment (johnstcn) [reply cli/create.go:198]: Ah, I misunderstood. 👍

**Review by johnstcn — APPROVED** [review]

**Review by mafredri — COMMENTED** [review]
Awesome feature! 🤩

Noticed a few small nits but all in all this looks good. 👍🏻

File: cli/create.go [hunk 2]
```diff
@@ -189,6 +191,11 @@ func (r *RootCmd) create() *clibase.Cmd {
 			Description: "Specify a duration after which the workspace should shut down (e.g. 8h).",
 			Value:       clibase.DurationOf(&stopAfter),
 		},
+		clibase.Option{
+			Flag:        "build-options",
```
[the remark this answers is not in the document]
Comment (mtojek) [reply cli/create.go:195]: Yeah, that was my feeling too 👍 but I would not like to expand to too many extra flags. @bpmct what is you advice here?

File: cli/start.go [hunk 3]
```diff
@@ -45,3 +69,49 @@ func (r *RootCmd) start() *clibase.Cmd {
...
+
+	richParameters := make([]codersdk.WorkspaceBuildParameter, 0)
+	if !args.BuildOptions {
+		return &buildParameters{
+			richParameters: richParameters,
+		}, nil
+	}
+
+	for _, templateVersionParameter := range templateVersionParameters {
+		if !templateVersionParameter.Ephemeral || !templateVersionParameter.Mutable {
```
[the remark this answers is not in the document]
Comment (mtojek) [reply cli/start.go:99]: > So we don't want to prompt for mutable parameters here?

The idea is to `start` a potentially stopped workspace with last parameter values.

> Or can a parameter be both ephemeral and immutable?

Oh, you may be right! This mutability condition should be guaranteed by [schema](https://github.com/coder/terraform-provider-coder/blob/main/provider/parameter.go#L134). I will iterate on this to make sure that I don't miss any edge case.

**Review by bpmct — APPROVED** [review]

File: cli/create.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mafredri) [reply cli/create.go:195]: Agreed, I support keeping this as a boolean flag for the time being. The proposed flags are proposals that simply replace the current one. In the future people may want to be able to script this behavior, so passing key=values may become relevant then.

[the remark this answers is not in the document]
Comment (mtojek) [reply cli/create.go:195]: I will leave it as is for now.

File: cli/restart.go [hunk 4]
```diff
@@ -20,21 +22,39 @@ func (r *RootCmd) restart() *clibase.Cmd {
 			r.InitClient(client),
 		),
 		Options: clibase.OptionSet{
+			{
+				Flag:        "build-options",
```
[the remark this answers is not in the document]
Comment (mtojek) [reply cli/restart.go]: Fixed, I introduced `workspaceParameterFlags`.

File: cli/start.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mtojek) [reply cli/start.go]: Fixed

File: cli/restart.go [hunk 5]
```diff
@@ -20,21 +22,39 @@ func (r *RootCmd) restart() *clibase.Cmd {
...
+			if err != nil {
+				return nil
+			}
+
+			buildParams, err := prepStartWorkspace(inv, client, prepStartWorkspaceArgs{
+				Template:     template,
+				BuildOptions: buildOptions,
+			})
+			if err != nil {
+				return nil
```
[the remark this answers is not in the document]
Comment (mtojek) [reply cli/restart.go]: Not sure what is the root cause of this problem, but fixed in a couple of places. Thank you for your diligence!

File: cli/create_test.go [hunk 6]
```diff
@@ -300,6 +305,46 @@ func TestCreateWithRichParameters(t *testing.T) {
...
+		for i := 0; i < len(matches); i += 2 {
+			match := matches[i]
+			value := matches[i+1]
+			pty.ExpectMatch(match)
+
+			if value != "" {
+				pty.WriteLine(value)
+			}
+		}
+		<-doneChan
```
[the remark this answers is not in the document]
Comment (mtojek) [reply cli/create_test.go:347]: Done

**Review by mafredri — APPROVED** [review]

**Pull request merged**